torchbiggraph/converters/importers.py

```diff
@@ -0,0 +1,388 @@
+#!/usr/bin/env python3
+
+# Copyright (c) Facebook, Inc. and its affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE.txt file in the root directory of this source tree.
+
+from abc import ABC, abstractmethod
+import random
+from contextlib import ExitStack
+from pathlib import Path
+from typing import Any, Counter, Dict, Iterable, List, Optional, Tuple, Union
+
+import torch
+
+from torchbiggraph.config import (
+    EntitySchema,
+    RelationSchema,
+)
+from torchbiggraph.converters.dictionary import Dictionary
+from torchbiggraph.edgelist import EdgeList
+from torchbiggraph.entitylist import EntityList
+from torchbiggraph.graph_storages import (
+    AbstractEdgeAppender,
+    AbstractEdgeStorage,
+    AbstractEntityStorage,
+    AbstractRelationTypeStorage,
+    EDGE_STORAGES,
+    ENTITY_STORAGES,
+    RELATION_TYPE_STORAGES,
+)
+
+
+class EdgelistReader(ABC):
+    @abstractmethod
+    def read(self, path: Path) -> Iterable[Tuple[str, str, Optional[str]]]:
+        """Read rows from a path. Returns (lhs, rhs, rel)."""
+        pass
+
+
+class TSVEdgelistReader(EdgelistReader):
+    def __init__(
+            self,
+            lhs_col: int,
+            rhs_col: int,
+            rel_col: int,
+    ):
+        self.lhs_col, self.rhs_col, self.rel_col = lhs_col, rhs_col, rel_col
+
+    def read(self, path: Path):
+        with path.open("rt") as tf:
+            for line_num, line in enumerate(tf, start=1):
+                try:
+                    words = line.split()
+                    lhs_word = words[self.lhs_col]
+                    rhs_word = words[self.rhs_col]
+                    rel_word = words[self.rel_col] if self.rel_col is not None else None
+                    yield lhs_word, rhs_word, rel_word
+                except IndexError:
+                    raise RuntimeError(
+                        f"Line {line_num} of {path} has only {len(words)} words"
+                    ) from None
+
+
+class ParquetEdgelistReader(EdgelistReader):
+    def __init__(
+            self,
+            lhs_col: str,
+            rhs_col: str,
+            rel_col: Optional[str],
+    ):
+        """Reads edgelists from a Parquet file.
+
+        col arguments can either be the column name or the offset of the col.
+        """
+        self.lhs_col, self.rhs_col, self.rel_col = lhs_col, rhs_col, rel_col
+
+    def read(self, path: Path):
+        try:
+            import parquet
+        except ImportError as e:
+            raise ImportError(f"{e}. HINT: You can install Parquet by running "
+                              "'pip install parquet'")
+
+        with path.open() as tf:
```

Comment:
Nit: pass `"rt"` as argument to `open`, for consistency and because explicit is better than implicit (cit. [The Zen of Python](https://www.python.org/dev/peps/pep-0020/))